src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -0,0 +1,183 @@
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, newValue -> timeout = newValue);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> indexingBulkSize = newValue);
+    }
+
+    @Override
+    protected void doExecute(final Task task, final PutDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        try {
+            Datasource jobParameter = Datasource.Builder.build(request);
+            IndexRequest indexRequest = new IndexRequest().index(DatasourceExtension.JOB_INDEX_NAME)
+                .id(jobParameter.getId())
+                .source(jobParameter.toXContent(JsonXContent.contentBuilder(), null))
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .opType(DocWriteRequest.OpType.CREATE);
+            client.index(indexRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(final IndexResponse indexResponse) {
+                    threadPool.generic().submit(() -> {
+                        try {
+                            createDatasource(jobParameter);
+                        } catch (Exception e) {
+                            log.error("Failed to create datasource for {}", jobParameter.getId(), e);
+                            jobParameter.getUpdateStats().setLastFailedAt(Instant.now());
+                            jobParameter.setState(DatasourceState.FAILED);
+                            try {
+                                DatasourceHelper.updateDatasource(client, jobParameter, timeout);
+                            } catch (Exception ex) {
+                                log.error("Failed to mark datasource state as FAILED for {}", jobParameter.getId(), ex);
+                            }
+                        }
+                    });
+                    listener.onResponse(new AcknowledgedResponse(true));
+                }
+
+                @Override
+                public void onFailure(final Exception e) {
+                    if (e instanceof VersionConflictEngineException) {
+                        log.info("Datasource already exists {}", request.getDatasourceName(), e);
+                        listener.onFailure(new ResourceAlreadyExistsException("Datasource already exists"));
+                    } else {
+                        log.error("Failed to create a datasource {}", request.getDatasourceName(), e);
+                        listener.onFailure(new OpenSearchException("Failed to create a datasource"));
+                    }
+                }
+            });
+        } catch (Exception e) {
+            log.error("Error occurred while creating datasource {}", request.getDatasourceName(), e);
+            listener.onFailure(new OpenSearchException("Failed to create a datasource"));
```

Comment:
why we are creating new exception here? and not passing the exception caught to the failure handler?

Reply:
But why cannot you rap the exception you caught in the OpenSearchException?

something like
```
new OpenSearchException("Failed to create a datasource", e);
```

Also, the error message you are sending to listener is very vague and doesn't provide details around what the customer should do, if this exception is getting translated to customer. 

Also, if in the listener we are log the exception I see that for 1 error we are logging the exception twice, why would we want to do that?

Reply:
Discussed offline. Will wrap with Geospatial custom exception

Reply:
Thought about it again. I think my code is valid. I don't want to expose the implementation detail to the caller. The listener will handle when it fails accordingly but it doesn't need to know the origin of the exception which is implementation detail.

For the message, the reason why it is vague is because we cannot provide more specific guideline in this case. The failure reason can be many different things from hardware failure to code bug. This exception is unexpected one. I cannot send a message something like "failed during indexing job parameter of datasource in an index" because it only reveals implementation details and still does not provide what the customer should do.

Reply:
> The listener will handle when it fails accordingly but it doesn't need to know the origin of the exception which is implementation detail.

How do we think that this happening?

Reply:
> For the message, the reason why it is vague is because we cannot provide more specific guideline in this case.

Let's try to think like this, if you are a customer who sees this exception message what you will do or a developer who sees this log and want to know what do? how you will proceed from there? I would recommend thinking in those lines. As this is so very vague.

Reply:
> How do we think that this happening?
It is the method's decision to reveal or not the origin of exception. If it thinks it will be useful to the caller, it will include them otherwise it doesn't have to. This is more or less case by case. For this specific line of code, I wanted to hide the detail as the caller will be a customer who don't need to know about the origin of the exception. Like IOException or NullPointerException on which they can do nothing about it. That is not what we want and we want to pass all those information to a customer?

> Let's try to think like this, if you are a customer who sees this exception message what you will do or a developer who sees this log and want to know what do? how you will proceed from there? I would recommend thinking in those lines. As this is so very vague.

For a customer all they can do is contact system admin to see what is the issue as it will be mostly hardware issue or bug in a code. For a developer, they need to look look at the log to figure it out. I agree that it is vague but the other option is just re-throwing the exception as it is.

Reply:
Checked other code in OpenSearch. It seems like it doesn't wrap the internal exception. Will follow same.